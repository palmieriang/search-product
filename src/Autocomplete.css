.search-container {
  width: 30rem;
}

.search-box {
  width: 100%;
  padding: 1rem;
  font-size: 1rem;
  border: 2px solid #e4e4e4;
}

.search-box:focus {
  outline: 1px solid #16a2ff;
}

.suggestions {
  margin: 0;
  padding-left: 0;
}

.suggestions li {
  background-color: white;
  border: 1px solid #e4e4e4;
  border-top-width: 0;
  list-style: none;
  padding: 1rem;
}

.suggestions .suggestion-active,
.suggestions li:hover {
  background-color: #f8f8f8;
  cursor: pointer;
}
