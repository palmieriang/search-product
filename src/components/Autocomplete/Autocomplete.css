.search-container {
  width: 30rem;
}

.search-box {
  width: 100%;
  padding: 1rem;
  font-size: 1rem;
  border: 2px solid #e4e4e4;
  border-radius: 4px;
}

.search-box:focus {
  outline: 1px solid #16a2ff;
}
