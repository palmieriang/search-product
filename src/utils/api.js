export const fetchSuggestions = (searchTerm) => {
  return fetch(
    "https://asoscarewebinterviewstub.azurewebsites.net/api/search",
    {
      method: "POST",
      body: JSON.stringify({
        q: searchTerm,
      }),
    }
  ).then((res) => res.json());
};

export const fetchProductDetail = (id) => {
  return fetch(
    "https://asoscarewebinterviewstub.azurewebsites.net/api/product",
    {
      method: "POST",
      body: JSON.stringify({
        id,
      }),
    }
  ).then((res) => res.json());
};
