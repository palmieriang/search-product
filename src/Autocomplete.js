import React, { useState, useEffect, useCallback, useRef } from "react";
import PropTypes from "prop-types";
import { debounce } from "lodash";

import { fetchSuggestions } from "./utils/api";

import "./Autocomplete.css";

const Autocomplete = ({ onSelect }) => {
  const [searchTerm, setSearchTerm] = useState('');
  const [suggestions, setSuggestions] = useState(null);
  const [fetchError, setFetchError] = useState(false);
  const [isLoading, setIsLoading] = useState(false);
  const [noMatches, setNoMatches] = useState(false);
  const inputEl = useRef(null);

  const fetch = (term) => {
    setIsLoading(true);
    fetchSuggestions(term)
      .then(
        (_suggestions) => {
          if (_suggestions.length > 0) {
            const showedSuggestions = _suggestions.slice(0, 10);
            setNoMatches(false);
            setIsLoading(false);
            setSuggestions(showedSuggestions);
          } else {
            setNoMatches(true);
            setIsLoading(false);
          }
        },
        (error) => {
          setIsLoading(false);
          setFetchError(true);
        }
      )
  }
  const debouncedFetch = useCallback(debounce(fetch, 500), []);

  useEffect(() => {
    if (searchTerm) {
      if (fetchError) {
        setFetchError(false);
      }
      debouncedFetch(searchTerm);
    } else {
      setSuggestions(null);
    }
  }, [debouncedFetch, fetchError, searchTerm]);

  const handleOnChange = useCallback((event) => {
    setSearchTerm(event.target.value);
  }, [setSearchTerm]);

  const handleOnKeyDown = useCallback((event) => {
    if (event.keyCode === 13) { // User pressed the enter key
      if (searchTerm && !suggestions) {
        event.preventDefault();
      }
      if (event.target.dataset.id) {
        onSelect(event.target.dataset.id);
        setSearchTerm('');
        inputEl.current.focus();
      } else if (suggestions?.length) {
        onSelect(suggestions[0].id);
        setSearchTerm('');
        inputEl.current.focus();
      }
    }
  }, [onSelect, searchTerm, suggestions]);

  const handleChooseProduct = useCallback((event) => {
    const index = parseInt(event.target.dataset.index, 10);
    onSelect(suggestions[index].id);
    setSearchTerm('');
    inputEl.current.focus();
  }, [onSelect, suggestions]);

  return (
    <div className="search-container">
      <label htmlFor="search-box" className="visually-hidden">Search for a product or brand</label>
      <input
        ref={inputEl}
        id="search-box"
        type="search"
        value={searchTerm}
        className="search-box"
        placeholder="Search for a product"
        onChange={handleOnChange}
        onKeyDown={handleOnKeyDown}
      />

      {isLoading && !suggestions && <p className="search-message">Loading...</p>}
      {noMatches && <p className="search-message">No matches</p>}
      {fetchError && <p className="error-message">Something went wrong</p>}

      {suggestions?.length > 0 && !noMatches && (
        <ul className="suggestions">
          {suggestions.map((suggestion, index) => {

            return (
              <li
                data-index={index}
                key={suggestion.id}
              >
                <button
                  className="suggestion-button"
                  data-index={index}
                  data-id={suggestion.id}
                  onClick={handleChooseProduct}
                  onKeyDown={handleOnKeyDown}
                >
                  {suggestion.name}
                </button>
              </li>
            )
          })}
        </ul>
      )}
    </div>
  );
}

Autocomplete.propTypes = {
  onSelect: PropTypes.func,
};

export default Autocomplete;
