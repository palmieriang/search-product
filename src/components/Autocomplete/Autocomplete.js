import React, { useEffect, useState } from "react";

import { fetchSuggestions } from "../../utils/api";

import "./Autocomplete.css";

const Autocomplete = () => {
  const [searchTerm, setSearchTerm] = useState("");
  const [suggestions, setSuggestions] = useState([]);

  useEffect(() => {
    fetchSuggestions(searchTerm).then((_suggestions) =>
      setSuggestions(_suggestions)
    );
  }, [searchTerm]);

  return (
    <div className="search-container">
      <input
        type="text"
        value={searchTerm}
        className="search-box"
        placeholder="Search for a product"
        onChange={(e) => setSearchTerm(e.target.value)}
      />
      {/* TODO: render search suggestions */}
    </div>
  );
}

export default Autocomplete;
