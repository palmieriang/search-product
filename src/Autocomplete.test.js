import React from "react";
import {render, screen, fireEvent, waitFor} from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import Autocomplete from "./Autocomplete";
import { fetchSuggestions } from "./utils/api";

jest.mock("./utils/api");

describe("Autocomplete", () => {
  beforeEach(() => {
    fetchSuggestions.mockClear()
  });

  it("renders correctly", () => {
    const { asFragment } = render(<Autocomplete />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("should not fetch if the user has not entered a search query", () => {
    render(<Autocomplete />);

    expect(fetchSuggestions).not.toHaveBeenCalled();
  });

  it("should fetch if the user has entered a search query", async () => {
    fetchSuggestions.mockResolvedValue([{id: 1, name: "shirt"}, {id: 2, name: "coat"}]);

    render(<Autocomplete />);

    const input = screen.getByLabelText(/Search/);
    const query = "adidas";

    fireEvent.change(input, { target: { value: query } });

    await waitFor(() => expect(fetchSuggestions).toHaveBeenCalledWith(query));
  });

  it("should show suggestions", async () => {
    fetchSuggestions.mockResolvedValueOnce([{id: 1, name: "shirt"}]);

    render(<Autocomplete />);

    fireEvent.change(screen.getByLabelText(/Search/), { target: { value: "shirt" } });

    await waitFor(() => expect(screen.getByText(/shirt/)).toBeInTheDocument());
  });

  it("should only show the top 10 suggestions at a time", async () => {
    const suggestions = [
      {id: "1", name: "shirt 1"},
      {id: "2", name: "shirt 2"},
      {id: "3", name: "shirt 3"},
      {id: "4", name: "shirt 4"},
      {id: "5", name: "shirt 5"},
      {id: "6", name: "shirt 6"},
      {id: "7", name: "shirt 7"},
      {id: "8", name: "shirt 8"},
      {id: "9", name: "shirt 9"},
      {id: "10", name: "shirt 10"},
      {id: "11", name: "shirt 11"},
    ];
    fetchSuggestions.mockResolvedValueOnce(suggestions);

    render(<Autocomplete />);

    fireEvent.change(screen.getByLabelText(/Search/), { target: { value: "query" } });

    await waitFor(() => expect(screen.getAllByText(/shirt/)).toHaveLength(10));
  });

  it("should not request for suggestions on every keystroke", async () => {
    const suggestions = [
      {id: "1", name: "shirt 1"},
    ];
    fetchSuggestions.mockResolvedValueOnce(suggestions);

    render(<Autocomplete />);

    const input = screen.getByLabelText(/Search/);

    userEvent.type(input, "queryasdasdasdasdassadsadadsdsadsa");

    await waitFor(() => expect(fetchSuggestions).toHaveBeenCalledTimes(1));
  })
});
