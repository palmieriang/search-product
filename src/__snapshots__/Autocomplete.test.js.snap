// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Autocomplete renders correctly 1`] = `
<div
  className="search-container"
>
  <label
    className="visually-hidden"
    htmlFor="search-box"
  >
    Search for a product or brand
  </label>
  <input
    className="search-box"
    id="search-box"
    onChange={[Function]}
    onKeyDown={[Function]}
    placeholder="Search for a product"
    type="search"
    value=""
  />
</div>
`;
